QUIC conceptual docs (#34601)
dotnet/docs: docs/fundamentals/networking/quic/quic-overview.md

---
title: QUIC support in .NET
description: Learn about the support for QUIC protocol in .NET.
ms.date: 03/15/2023
helpviewer_keywords:
  - "protocols, QUIC"
  - "sending data, QUIC"
  - "QUIC"
  - "receiving data, QUIC"
  - "application protocols, QUIC"
  - "Internet, QUIC"
---

> # QUIC

Author: Done.

> QUIC is a new, transport layer protocol. It has been recently standardized in [RFC 9000](https://www.rfc-editor.org/rfc/rfc9000). It uses UDP as an underlying protocol and it's inherently secure as it mandates TLS 1.3 usage, see [RFC 9001](https://www.rfc-editor.org/rfc/rfc9001). Another interesting difference from well-known transport protocols such as TCP and UDP is that it has stream multiplexing built-in on the transport layer. This allows having multiple, concurrent, independent data streams that do not affect each other.

Author: Done

> QUIC itself doesn't define any semantics for the exchanged data as it's a transport protocol. It's rather used in application layer protocols, for example in [HTTP/3](https://www.rfc-editor.org/rfc/rfc9114) or in [SMB over QUIC](https://learn.microsoft.com/windows-server/storage/file-server/smb-over-quic). It can also be used for any custom defined protocol.

Author: Done

> The protocol offers many advantages over TCP with TLS. For instance, faster connection establishment as it doesn't require as many round trips as TCP with TLS on top. Or avoidance of head-of-line blocking problem where one lost packet doesn't block data of all the other streams. On the other hand, there are disadvantages that come with using QUIC. As it is a new protocol, its adoption is still growing and is limited. Apart from that, QUIC traffic might be even blocked by some networking components.

Author: Done.

## QUIC in .NET

> We introduced QUIC implementation in .NET 5 in `System.Net.Quic` library. However, up until .NET 7.0 the library was strictly internal and served only for our own implementation of HTTP/3. With .NET 7, we made the library public and we exposed its APIs.

Author: Done.

> [!NOTE]

> >

Author: Done

> > For now, the APIs are published with [preview feature](https://github.com/dotnet/designs/blob/main/accepted/2021/preview-features/preview-features.md).

Author: Done.

---
title: QUIC support in .NET
description: Learn about the support for QUIC protocol in .NET.
ms.date: 03/15/2023
helpviewer_keywords:
  - "protocols, QUIC"
  - "sending data, QUIC"
  - "QUIC"
  - "receiving data, QUIC"
  - "application protocols, QUIC"
  - "Internet, QUIC"
---
# QUIC

> QUIC is a new, transport layer protocol. It has been recently standardized in [RFC 9000](https://www.rfc-editor.org/rfc/rfc9000). It uses UDP as an underlying protocol and it's inherently secure as it mandates TLS 1.3 usage, see [RFC 9001](https://www.rfc-editor.org/rfc/rfc9001). Another interesting difference from well-known transport protocols such as TCP and UDP is that it has stream multiplexing built-in on the transport layer. This allows having multiple, concurrent, independent data streams that do not affect each other.

Reviewer: Just a thought, maybe it will be better to reword as "QUIC is a transport layer protocol standardized in [RFC 9000](https://www.rfc-editor.org/rfc/rfc9000)." -- given it's a doc page that will potentially stay for a long time when "recently" might no longer apply 😄
(I see the same "new" wording in H/3 doc as well, so if you'd agree with my proposal, I think it might be better to remove "new" from there as well)
WDYT?

The sample usage of `QuicConnection`:

```C#
using System.Net.Quic;

// First, check if QUIC is supported.
if (!QuicConnection.IsSupported)
{
    Console.WriteLine("QUIC is not supported, check for presence of libmsquic and support of TLS 1.3.");
    return;
}

// This represents the minimal configuration necessary to open a connection.
var clientConnectionOptions = new QuicClientConnectionOptions()
{
    // End point of the server to connect to.
    RemoteEndPoint = listener.LocalEndPoint,

    // Used to abort stream if it's not properly closed by the user.
    // See https://www.rfc-editor.org/rfc/rfc9000#section-20.2
    DefaultStreamErrorCode = 0x0A, // Protocol-dependent error code.

    // Used to close the connection if it's not done by the user.
    // See https://www.rfc-editor.org/rfc/rfc9000#section-20.2
    DefaultCloseErrorCode = 0x0B, // Protocol-dependent error code.

    // Optionally set limits for inbound streams.
    MaxInboundUnidirectionalStreams = 10,
    MaxInboundBidirectionalStreams = 100,

    // Same options as for client side SslStream.

>     ClientAuthenticationOptions = new SslClientAuthenticationOptions()

Reviewer 2: ```suggestion
    ClientAuthenticationOptions = new SslClientAuthenticationOptions
```

    {
        // List of supported application protocols.
        ApplicationProtocols = new List<SslApplicationProtocol>() { "protocol-name" }
    }
};

// Initialize, configure and connect to the server.
var connection = await QuicConnection.ConnectAsync(clientConnectionOptions);

Console.WriteLine($"Connected {connection.LocalEndPoint} --> {connection.RemoteEndPoint}");

// Open a bidirectional (can both read and write) outbound stream.
var outgoingStream = await connection.OpenOutboundStreamAsync(QuicStreamType.Bidirectional);

// Work with the outgoing stream ...

// To accept any stream on a client connection, at least one of MaxInboundBidirectionalStreams or MaxInboundUnidirectionalStreams of QuicConnectionOptions must be set.
while (isRunning)
{
    // Accept an inbound stream.
    var incomingStream = await connection.AcceptInboundStreamAsync();

    // Work with the incoming stream ...
}

// Close the connection with the custom code.
await connection.CloseAsync(0x0C);

// Dispose the connection.
await connection.DisposeAsync();
```

> More details about how this class was designed can be found in the `QuicConnection` API Proposal ([dotnet/runtime#68902](https://github.com/dotnet/runtime/issues/68902)).

Reviewer 2: ```suggestion
For more information about how the `QuicConnection` was designed, see the [API proposal](https://github.com/dotnet/runtime/issues/68902).
```

> ### QuicStream

Reviewer 2: ```suggestion
### `QuicStream`
```

Another particularity of QUIC stream is ability to explicitly close the writing side in the middle of work with the stream, see <xref:System.Net.Quic.QuicStream.CompleteWrites> or <xref:System.Net.Quic.QuicStream.WriteAsync(System.ReadOnlyMemory{System.Byte},System.Boolean,System.Threading.CancellationToken)> overload with `completeWrites` argument. Closing of the writing side lets the peer know that no more data will arrive, yet the peer still can continue sending (in case of a bidirectional stream). This is useful in scenarios like HTTP request/response exchange when the client sends the request and closes the writing side to let the server know that this is the end of the request content. Server is still able to send the response after that, but knows that no more data will arrive from the client. And for erroneous cases, either writing or reading side of the stream can be aborted, see <xref:System.Net.Quic.QuicStream.Abort(System.Net.Quic.QuicAbortDirection,System.Int64)>. The behavior of the individual methods for each stream type is summarized in the following table (note that both client and server can open and accept streams):

> | Method     | Peer opening stream  | Peer accepting stream  |
> | -          | -       | -       |
> | `CanRead`  | _bidirectional_: `true`<br/> _unidirectional_: `false` | `true`  |
> | `CanWrite` | `true`  | _bidirectional_: `true`<br/> _unidirectional_: `false` |
> | `ReadAsync` | _bidirectional_: reads data<br/> _unidirectional_: `InvalidOperationException` | reads data |
> | `WriteAsync` | sends data => peer read returns the data | _bidirectional_: sends data => peer read returns the data<br/> _unidirectional_: `InvalidOperationException`  |
> | `CompleteWrites` | closes writing side => peer read returns 0 | _bidirectional_: closes writing side => peer read returns 0<br/> _unidirectional_: no-op |
> | `Abort(QuicAbortDirection.Read)` | _bidirectional_: [STOP_SENDING](https://www.rfc-editor.org/rfc/rfc9000#section-19.5) => peer write throws `QuicException(QuicError.OperationAborted)`<br/> _unidirectional_: no-op | [STOP_SENDING](https://www.rfc-editor.org/rfc/rfc9000#section-19.5) => peer write throws `QuicException(QuicError.OperationAborted)`|
> | `Abort(QuicAbortDirection.Write)` | [RESET_STREAM](https://www.rfc-editor.org/rfc/rfc9000#section-19.4) => peer read throws `QuicException(QuicError.OperationAborted)` | _bidirectional_: [RESET_STREAM](https://www.rfc-editor.org/rfc/rfc9000#section-19.4) => peer read throws `QuicException(QuicError.OperationAborted)`<br/> _unidirectional_: no-op |

Reviewer 2: ```suggestion
| Method | Peer opening stream | Peer accepting stream |
|---|---|---|
| `CanRead` | _bidirectional_: `true`<br/> _unidirectional_: `false` | `true` |
| `CanWrite` | `true` | _bidirectional_: `true`<br/> _unidirectional_: `false` |
| `ReadAsync` | _bidirectional_: reads data<br/> _unidirectional_: `InvalidOperationException` | reads data |
| `WriteAsync` | sends data => peer read returns the data | _bidirectional_: sends data => peer read returns the data<br/> _unidirectional_: `InvalidOperationException` |
| `CompleteWrites` | closes writing side => peer read returns 0 | _bidirectional_: closes writing side => peer read returns 0<br/> _unidirectional_: no-op |
| `Abort(QuicAbortDirection.Read)` | _bidirectional_: [STOP_SENDING](https://www.rfc-editor.org/rfc/rfc9000#section-19.5) => peer write throws `QuicException(QuicError.OperationAborted)`<br/> _unidirectional_: no-op | [STOP_SENDING](https://www.rfc-editor.org/rfc/rfc9000#section-19.5) => peer write throws `QuicException(QuicError.OperationAborted)` |
| `Abort(QuicAbortDirection.Write)` | [RESET_STREAM](https://www.rfc-editor.org/rfc/rfc9000#section-19.4) => peer read throws `QuicException(QuicError.OperationAborted)` | _bidirectional_: [RESET_STREAM](https://www.rfc-editor.org/rfc/rfc9000#section-19.4) => peer read throws `QuicException(QuicError.OperationAborted)`<br/> _unidirectional_: no-op |
```

Finally, when the work with the stream is done, it needs to be disposed with <xref:System.Net.Quic.QuicStream.DisposeAsync>. The dispose will make sure that both reading and/or writing side - depending on the stream type - is closed. If stream hasn't been properly read till the end, dispose will issue an equivalent of `Abort(QuicAbortDirection.Read)`. However, if stream writing side hasn't been closed, it will be gracefully closed as it would be with `CompleteWrites`. The reason for this difference is to make sure that scenarios working with an ordinary `Stream` behave as expected and lead to a successful path. Consider the following example:

> ```C#

Reviewer 2: ```suggestion
```csharp
```

// Work done with all different types of streams.

> async Task WorkWithStream(Stream stream)

Reviewer 2: ```suggestion
async Task WorkWithStreamAsync(Stream stream)
```

{
    // This will dispose the stream at the end of the scope.
    await using (stream)
    {
        // Simple echo, read data and send them back.
        byte[] buffer = new byte[1024];
        int count = 0;
        // The loop stops when read returns 0 bytes as is common for all streams.
        while ((count = await stream.ReadAsync(buffer)) > 0)
        {
            await stream.WriteAsync(buffer.AsMemory(0, count));
        }
    }
}

// Open a QuicStream and pass to the common method.
var quicStream = await connection.OpenOutboundStreamAsync(QuicStreamType.Bidirectional);
await WorkWithStream(quicStream);
```

The sample usage of `QuicStream` in client scenario:

> ```C#

Reviewer 2: ```suggestion
```csharp
```

// Consider connection from the connection example, open a bidirectional stream.
await using var stream = await connection.OpenStreamAsync(QuicStreamType.Bidirectional, cancellationToken);

// Send some data.
await stream.WriteAsync(data, cancellationToken);
await stream.WriteAsync(data, cancellationToken);

// End the writing-side together with the last data.
await stream.WriteAsync(data, endStream: true, cancellationToken);
// Or separately.
stream.CompleteWrites();

// Read data until the end of stream.
while (await stream.ReadAsync(buffer, cancellationToken) > 0)
{
    // Handle buffer data...
}

// DisposeAsync called by await using at the top.
```

And the sample usage of `QuicStream` in server scenario:

> ```C#

Reviewer 2: ```suggestion
```csharp
```

### `QuicListener`

<xref:System.Net.Quic.QuicListener> represents a server side class that accepts incoming connections from the clients. The listener is constructed and started with a static method <xref:System.Net.Quic.QuicListener.ListenAsync(System.Net.Quic.QuicListenerOptions,System.Threading.CancellationToken)>. The method accepts an instance of <xref:System.Net.Quic.QuicListenerOptions> class with all the settings necessary to start the listener and accept incoming connections. After that, listener is ready to hand out connections via <xref:System.Net.Quic.QuicListener.AcceptConnectionAsync(System.Threading.CancellationToken)>. Connections returned by this method are always fully connected, meaning that the TLS handshake is finished and the connection is ready to be used. Finally, to stop listening and release all resources, <xref:System.Net.Quic.QuicListener.DisposeAsync> must be called.

Consider the following `QuicListener` example code:

> 

Reviewer 2: ```suggestion
```

When the work with the connection is done, it needs to be closed and disposed. QUIC protocol mandates using an application layer code for immediate closure, see [RFC 9000 Section 10.2](https://www.rfc-editor.org/rfc/rfc9000#section-10.2). For that, <xref:System.Net.Quic.QuicConnection.CloseAsync(System.Int64,System.Threading.CancellationToken)> with application layer code can be called or if not, <xref:System.Net.Quic.QuicConnection.DisposeAsync> will use the code provided in <xref:System.Net.Quic.QuicConnectionOptions.DefaultCloseErrorCode>. Either way, <xref:System.Net.Quic.QuicConnection.DisposeAsync> must be called at the end of the work with the connection to fully release all the associated resources.

Consider the following `QuicConnection` example code:

> 

Reviewer 2: ```suggestion
```
